File: src/main/java/yuma140902/yumalib/loot/MobDropHandler.java
package yuma140902.yumalib.loot;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.item.ItemStack;
import net.minecraftforge.event.entity.living.LivingDropsEvent;
import yuma140902.yumalib.api.loot.IMobLootInfo;

public class MobDropHandler {
	
	public static MobDropHandler INSTANCE = new MobDropHandler();
	
	private MobDropHandler() {}
	
	private List<IMobLootInfo> registry = new ArrayList<IMobLootInfo>();
	
	public void register(IMobLootInfo info) {
		if(info != null && info.enabled()) registry.add(info);
	}
	
	public void onLivingDrop(LivingDropsEvent event) {
		Entity entity = event.entityLiving;
		Random rand = event.entityLiving.worldObj.rand;
		
		for(final IMobLootInfo info : registry) {
			if(info != null && info.matchEntityClass(entity) && info.extraValidate(event)) {
				event.drops.add(new EntityItem(entity.worldObj, entity.posX, entity.posY, entity.posZ, new ItemStack(info.item(), info.itemNum(rand), info.itemMeta())));
			}
		}
		
	}
}

File: src/main/scala/yuma140902/yumalib/loot/MobDropHandler.scala
package yuma140902.yumalib.loot

import yuma140902.yumalib.api.loot.MobLoot

import net.minecraft.entity.item.EntityItem
import net.minecraftforge.event.entity.living.LivingDropsEvent

import scala.collection.mutable

// TODO: yuma140902.yumalib.api.lootに移動する？
// TODO: registryとhandlerを分ける必要がある
object MobDropHandler {
  private val registry: mutable.ArrayStack[MobLoot] = new mutable.ArrayStack[MobLoot]()

  def register(mobLoot: MobLoot): Unit =
    if (mobLoot.enabled) registry.push(mobLoot)

  def onLivingDrop(event: LivingDropsEvent): Unit = {
    val entity = event.entityLiving
    val rand = event.entityLiving.worldObj.rand;

    registry.foreach(loot => {
      loot.newItemStack(rand, entity, event).map(stack => {
        event.drops.add(new EntityItem(entity.worldObj, entity.posX, entity.posY, entity.posZ, stack))
      })
    })
  }
}
